### ISSUE TITLE
Conditional formatting consolidation break priority ordering

### ISSUE THREAD

**Issue by nathanQiu1** [ClosedXML/ClosedXML#757]
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

0.92.0

**What is the current behavior?**

The sequence of "conditional formatting rules" has been changed after export Excel by using ClosedXML in my program

**What is the expected behavior or new feature?**

The sequence should not be changed in Excel template after export excel.

**Did this work in previous versions of our tool?  Which versions?**

No

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

Without a code sample, it is unlikely that your issue will get attention. Don't be lazy. Do the effort and assist the developers to reproduce your problem. Code samples should be [minimal complete and verifiable](https://stackoverflow.com/help/mcve). Sample spreadsheets should be attached whenever applicable. Remove sensitive information.

**Code to reproduce problem:**
```c#
public void Main()
{
      using (XLWorkbook workbook = new XLWorkbook(@"FAI.xlsm"))
   {
       workbook.Save();
   }
}
```
- [x] I attached a sample spreadsheet.  (You can drag files on to this issue)
[test4.xlsx](https://github.com/ClosedXML/ClosedXML/files/1839700/test4.xlsx)

**Comment by nathanQiu1** [comment]
This change is very subtle, not easy to find, only appears under a specific template. No excel tool was used.

**Comment by igitur** [comment]
So how was the template originally created?

**Comment by nathanQiu1** [comment]
The friend gave me, he said the same as creating an ordinary file (xlsx), the only difference is the "conditional formatting rules", which he manually edited. The test4 file was formed by my friend's template simplification.
If you have other needs, please notify me immediately. Thank you！

**Comment by igitur** [comment]
The file is still massively complicated. For example: 57 font styles are defined. 30 types of background fills, 19 types of borders, 1339 cell styles (combining fonts, backgrounds, borders, etc), 472 differential formats.

I'm sorry, but really need you to strip down the file to the simplest possible file that still reproduces the problem.

**Comment by igitur** [comment]
Start by deleting all the unwanted cells. I just need 1 cell to show the problem, not an entire column.

**Comment by igitur** [comment]
Unhide and delete the irrelevant sheets too.

**Comment by nathanQiu1** [comment]
I have tried to separate, but there is no problem after separation. I may
need more time to try again.
thank you!

2018-03-23 16:52 GMT+08:00 Francois Botha <notifications@github.com>:

**Comment by igitur** [comment]
It looks like #709 will solve this.

**Comment by Pankraty** [comment]
![image](https://user-images.githubusercontent.com/19576939/37836309-99672fe6-2ecb-11e8-901e-c840311678ed.png)
These 6 borders look suspicious for me. 
One of the fills is duplicated too:
![image](https://user-images.githubusercontent.com/19576939/37836516-14c6adec-2ecc-11e8-89f8-9d34dc293f5a.png)

@igitur  do you mind if I proceed with this on weekend?

**Comment by igitur** [comment]
@pankraty go ahead

**Comment by Pankraty** [comment]
One of the issues is caused by the fact that ClosedXML assumes that conditional formats are added in order of priorities, while in the sample file they follow in arbitrary order (starting from 521, then 518, then 519 and so on). To fix this I am going to add a ```Priority``` property to ```IXLConditionalFormat```. This does not fix the original problem but seems necessary anyway.

**Comment by Pankraty** [comment]
I need to check it, thanks for the reference

**Comment by igitur** [comment]
Also, I don't think the conditional format priority should be exposed on the public API. It's an internal matter related to loading and saving. In fact zip when saving i prefer to save them in the order of priority, meaning that the order of the XML elements match the order of the priority attributes.

**Comment by Pankraty** [comment]
I did not merge the latest develop branch and missed that the ordering has been fixed recently. Now I think the problem is in CF consolidation. 
When the consolidation is applied, CF rules affecting cell L1 are these:
![image](https://user-images.githubusercontent.com/19576939/37865539-62e5dfb8-2f97-11e8-9b49-0d080a8b498f.png)

If I turn consolidation off the set of CF rules looks like this:
![image](https://user-images.githubusercontent.com/19576939/37865514-f6bbe0a8-2f96-11e8-8fe0-51a55aa001d1.png)

Apparently, when CFs are consolidated they may be reordered causing priority changes. I guess now I'll be able to create a more simple example reproducing the issue.

**Comment by igitur** [comment]
Maybe @b0bi79 can help here. He did the conditional format consolidation part.

**Comment by Pankraty** [comment]
Reproduction:
[ConditionalFormattingConsolidation.xlsx](https://github.com/ClosedXML/ClosedXML/files/1844592/ConditionalFormattingConsolidation.xlsx)

**Comment by nathanQiu1** [comment]
thanks for @Pankraty！

**Comment by Pankraty** [comment]
I am afraid, this issue is not something that can be easily fixed. Conceptually, I see the following options.

1. Do not perform a consolidation of conditional formats at all. The easiest way to implement, but it will cause multiplication of conditional format rules and will be of course a step back.
2. Modify ```IXLConditionalFormat``` for it to store ```Ranges``` collection instead of a single ```Range```. With this, it will be possible to read original conditional formats (that may be multi-ranged) and save them in their original form. Also, it will still be possible to perform a consolidation of CF rules, but this time with respect to CF priorities. This, however, would be a breaking change, since users would need to specify a collection of ranges rather than a single range. To make an upgrade smooth we could preserve a ```Range``` property for a version or two, just marking it as ```[Obsolete]```.
3. Perform a consolidation of CF rules on saving only, combining identical rules into one rule with a few ranges (again, preserving priorities). Inconvenience is that user will see different sets of rules in Excel and in ClosedXML, having no possibility to control how they are split and combined. What is good, the public interface will remain untouched.

I vote for the option 2 :)

**Comment by nathanQiu1** [comment]
thanks，option 2 is the best decision.

**Comment by Pankraty** [comment]
If the conditional format was originally applied to multiple disjoined ranges, it will be split into several independent CFs on file load. On saving, these rules won't be consolidated, because consolidation only applies to neighbor ranges. So the problem of multiplied CF rules would still exist.
If CFs will store a collection of ranges, the consolidation may be not needed in most cases. And when it is needed, I would go through the collection of CFs, combining those that have the same formula and style until I reach the rule intersecting any of the captured ranges. This would stop a consolidation of this particular rule to prevent priorities disturbance.

**Comment by Pankraty** [comment]
I prepared changes addressing this issue in the branch https://github.com/Pankraty/ClosedXML/commits/CFConsolidate. It depends on #709 and #758, and not finished yet (5 tests fail) but the approach looks eligible.

**Comment by nathanQiu1** [comment]
Sorry to bother you. It has been a week. Is there a good solution now?

**Comment by Pankraty** [comment]
I have prepared a solution but I would not expect it to be released before 0.93. Although it is up to @igitur.

**Comment by nathanQiu1** [comment]
If you can, maybe. . Let me test it。:laughing:

**Comment by Pankraty** [comment]
No problem. You can check out the branch https://github.com/Pankraty/ClosedXML/commits/CFConsolidate and build it.

**Comment by nathanQiu1** [comment]
Is there a change in method? After I build it , it did not resolve.

**Comment by Pankraty** [comment]
Make sure you checked out the correct branch and referenced the right assembly (the one you've built, not some cached one, by accident). I tested on the example you provided so it has to be fixed.

**Comment by nathanQiu1** [comment]
ok ,I try it again。
thanks。

**Comment by nathanQiu1** [comment]
Sorry, there may still be some issues. Please check out my new attachment.
[test5.xlsx](https://github.com/ClosedXML/ClosedXML/files/1870599/test5.xlsx)

**Comment by igitur** [comment]
I suggest we park this until we address #767 .

**Comment by nathanQiu1** [comment]
@Pankraty    Very regrettable! After I build it, it can solve the above problem, but cannot solve the 'test5.xlsx' file.

**Comment by Pankraty** [comment]
@nathanQiu1 I think I found a flaw in my algorithm. Thanks for the example.

**Comment by Pankraty** [comment]
Could you please check the attached file? For me, it looks OK now.

[757_sample2_changed.xlsx](https://github.com/ClosedXML/ClosedXML/files/1880723/757_sample2_changed.xlsx)
.

**Comment by nathanQiu1** [comment]
@Pankraty  I think it's basically the same. 
thanks!

**Comment by nathanQiu1** [comment]
@Pankraty  After I build it, Sometimes this prompt appears: Attribute 'id' should have unique value. Its current value '7' duplicates with others. in /xdr:wsDr[1]/xdr:oneCellAnchor[3]/xdr:pic[1]/xdr:nvPicPr[1]/xdr:cNvPr[1] Attribute 'id' should have unique value. Its current value '4' duplicates with others.
But 0.92.0 did not。

**Comment by Pankraty** [comment]
Strange. This branch did not change anything related to pictures, AFAIR. If you could open another issue, attaching your file having this problem that would be great.

**Comment by nathanQiu1** [comment]
Hi @Pankraty ：
     This template also has a color change.
[test10GRR.xlsx](https://github.com/ClosedXML/ClosedXML/files/1889778/test10GRR.xlsx)

**Comment by nathanQiu1** [comment]
Sorry to bother you @Pankraty . Is there a good solution now?

**Comment by Pankraty** [comment]
The first example has no relation to conditional formats consolidation.
The second one is relevant. For expression-based conditional formats, I thought relative addresses are calculated starting from the first cell in ranges definition. But in fact, they are calculated starting from the top left cell of the envelope range, even if this cell is not included in CF itself.
Here is the illustration:
![image](https://user-images.githubusercontent.com/19576939/38694032-a4c14538-3e99-11e8-8a69-f80dfaf341ec.png)
Will create a test case for this.

**Comment by nathanQiu1** [comment]
thanks for @Pankraty  !  
'757_sample4_changed.xlsx' is OK  ，But this does not solve the problem of the conditional format. Please see my two attachments below。
[test15.xlsx](https://github.com/ClosedXML/ClosedXML/files/1906016/test15.xlsx)
[test16.xlsx](https://github.com/ClosedXML/ClosedXML/files/1906017/test16.xlsx)

**Comment by Pankraty** [comment]
Both examples have no relation to the issue subject (Conditional formatting consolidation break priority ordering). 
Here, for cell L1 rules 3 and 4 are applied. Rule 3 modifies its background and rule 4 modifies the font color (since "Stop If True" is not checked).
![image](https://user-images.githubusercontent.com/19576939/38719935-6cae88b0-3f05-11e8-8e3d-8205aab21cc8.png)
In the original file rule nr. 3 defined foreground color explicitly, but ClosedXML considers default foreground value as being not specified. This issue is addressed in #767.
As a workaround, you may specify black color instead of Automatic for CF foreground.
![image](https://user-images.githubusercontent.com/19576939/38720194-5412e804-3f06-11e8-9561-0ee17de0f52e.png)

**Comment by nathanQiu1** [comment]
thanks, Is there any way to avoid changing the background color of 'test15.xlsx'?

**Comment by nathanQiu1** [comment]
How did you do it? There is no automatic background color for the four conditional formats.

**Comment by Pankraty** [comment]
On top, there is your file, on the bottom - this file saved with ClosedXML with my yesterday fix. I didn't do anything else.

**Comment by nathanQiu1** [comment]
before:    
![image](https://user-images.githubusercontent.com/19723900/38725108-decde3e0-3f38-11e8-9ee5-15f9c8dacded.png)

now : 
![image](https://user-images.githubusercontent.com/19723900/38725137-f558f212-3f38-11e8-9e85-f3039673c989.png)
I updated your new code yesterday, but still, please look at the attachment. Is it correct?
[test15-1.xlsx](https://github.com/ClosedXML/ClosedXML/files/1906530/test15-1.xlsx)

**Comment by Pankraty** [comment]
Make sure you are checking out the correct branch and try a clean rebuild. That's the result I'm getting from your file:
![image](https://user-images.githubusercontent.com/19576939/38725421-557f0e36-3f18-11e8-950c-83e0f8de6d0e.png)

**Comment by nathanQiu1** [comment]
Ok, I re-check, thank you！
